pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,351 @@
+	// for deterministic ordering
+	toUnscheduleSorted := types.NewNodeSet(currentNodes...).ListNodes()
+	ds.logger.NoFields().Infof("Need to unschedule %d nodes", len(toUnscheduleSorted))
+
+	for _, node := range toUnscheduleSorted {
+		err := ds.unschedule(node)
+		if err != nil {
+			return util.Errorf("Error scheduling node: %v", err)
+		}
+	}
+	return nil
+}
+
+func (ds *daemonSet) schedule(node types.NodeName) error {
+	ds.logger.NoFields().Infof("Scheduling on %s", node)
+
+	// Will apply the following label on the key <labels.POD>/<node>/<ds.Manifest.ID()>:
+	// 	{ DSIDLabel : ds.ID() }
+	// eg node/127.0.0.1[daemon_set_id] := test_ds_id
+	// This is for indicating that this pod path belongs to this daemon set
+	id := MakePodLabelKey(node, ds.Manifest.ID())
+	err := ds.applicator.SetLabel(labels.POD, id, DSIDLabel, ds.ID().String())
+	if err != nil {
+		return util.Errorf("Error setting label: %v", err)
+	}
+
+	// Will set the following the value, ds.Manifest, to the following key:
+	// <kp.INTENT_TREE>/<node>/<ds.Manifest.ID()>
+	// eg intent/127.0.0.1/test = test_pod
+	_, err = ds.kpStore.SetPod(kp.INTENT_TREE, node, ds.Manifest)
+	if err != nil {
+		return util.Errorf("Error adding pod to intent tree: %v", err)
+	}
+	return nil
+}
+
+func (ds *daemonSet) unschedule(node types.NodeName) error {
+	ds.logger.NoFields().Infof("Unscheduling from %s", node)
+
+	// Will remove the following key:
+	// <kp.INTENT_TREE>/<node>/<ds.Manifest.ID()>
+	_, err := ds.kpStore.DeletePod(kp.INTENT_TREE, node, ds.Manifest.ID())
+	if err != nil {
+		return util.Errorf("Unable to delete pod from intent tree: %v", err)
+	}
+
+	// Will remove the following label on the key <labels.POD>/<node>/<ds.Manifest.ID()>: DSIDLabel
+	// This is for indicating that this pod path no longer belongs to this daemon set
+	id := MakePodLabelKey(node, ds.Manifest.ID())
+	err = ds.applicator.RemoveLabel(labels.POD, id, DSIDLabel)
+	if err != nil {
+		return util.Errorf("Error removing label: %v", err)
+	}
+
+	return nil
+}
+
+func (ds *daemonSet) CurrentPods() (PodLocations, error) {
+	// Changing DaemonSet.ID is not permitted, so as long as there is no uuid
+	// collision, this will always get the current pod path that this daemon set
+	// had scheduled on
+	selector := klabels.Everything().Add(DSIDLabel, klabels.EqualsOperator, []string{ds.ID().String()})
+
+	podMatches, err := ds.applicator.GetMatches(selector, labels.POD)
+	if err != nil {
+		return nil, util.Errorf("Unable to get matches on pod tree: %v", err)
+	}
+
+	result := make(PodLocations, len(podMatches))
+	for i, podMatch := range podMatches {
+		// ID will be something like <node>/<PodID>
+		node, podID, err := labels.NodeAndPodIDFromPodLabel(podMatch)
+		if err != nil {
+			return nil, err
+		}
+		result[i].Node = node
+		result[i].PodID = podID
+	}
+
+	return result, nil
+}
+
+func MakePodLabelKey(node types.NodeName, podID types.PodID) string {
```

Comment:
bonus points if you put this in `pkg/labels/consul_applicator.go` and make replication controllers use it as well